src/ui/SmartSnippet/SmartSnippet.ts

```diff
@@ -162,15 +165,100 @@ export class SmartSnippet extends Component {
   }
 
   private renderSourceTitle(title: string, clickUri: string) {
-    const element = $$('a', { className: SOURCE_TITLE_CLASSNAME, href: clickUri }).el;
-    element.innerText = title;
-    return element;
+    return this.renderLink(title, clickUri, SOURCE_TITLE_CLASSNAME);
   }
 
   private renderSourceUrl(url: string) {
-    const element = $$('a', { className: SOURCE_URL_CLASSNAME, href: url }).el;
-    element.innerText = url;
+    return this.renderLink(url, url, SOURCE_URL_CLASSNAME);
+  }
+
+  private renderLink(text: string, href: string, className: string) {
+    const element = $$('a', { className, href }).el as HTMLAnchorElement;
+    element.innerText = text;
+    this.enableAnalyticsOnLink(element, () => this.sendOpenSourceAnalytics());
     return element;
   }
+
+  private fixSnippetLink(link: HTMLAnchorElement, sourceUrl?: string) {
+    const staticLink = link.cloneNode(true) as HTMLAnchorElement;
+    if (sourceUrl) {
+      const href = link.getAttribute('href');
+      link.href = UrlUtils.getLinkDestination(sourceUrl, href);
+    }
+    this.enableAnalyticsOnLink(link, () => this.sendOpenContentLinkAnalytics(staticLink));
+  }
+
+  private enableAnalyticsOnLink(link: HTMLAnchorElement, sendAnalytics: () => Promise<any>) {
+    link.addEventListener('click', e => {
+      e.preventDefault();
+      this.openLink(link.href, e.ctrlKey, sendAnalytics);
+    });
+  }
+
+  private async openLink(href: string, newTab: boolean, sendAnalytics: () => Promise<any>) {
+    await Promise.race<any>([sendAnalytics(), Utils.resolveAfter(OPEN_LINK_TIMEOUT)]);
```

Comment:
Why is `race` and `await` needed here? Why not just log and forget?

Reply:
Without `await`ing analytics, I couldn't see the outgoing query in the network tab, which made me question whether it was sent at all. I added this `race` to attempt to wait for the query to be sent but stop waiting past `OPEN_LINK_TIMEOUT`ms.

I'll check with Wireshark whether the query is sent without this, and if so, I'll remove this `await`.

Reply:
Hmm, we should have some code leveraging the Beacon API to ensure that requests are sent before page unload. Maybe that could be of help?

Ref: https://developer.mozilla.org/en-US/docs/Web/API/Beacon_API

Reply:
The Beacon API sounds like a perfect fit, however, it isn't compatible with IE.

Reply:
You should not need any of this as this is already handled. If you look into the `AnalyticsEndpointCaller.ts` file, it's already using beacon API, with fallback when it's not available.

Your code should just be `logAnalytics()`, fire and forget, rest of the code downstream will handle it.

Beacon calls in Chrome console won't appear in XHR tab either, they appear in a different subtab.

Reply:
Ah, sweet! Then I'll remove this await.